README.md
# time-to-go

[![Build Status](https://travis-ci.org/tdsh/time-to-go.svg?branch=master)](https://travis-ci.org/tdsh/time-to-go)

Simple command to trigger an alarm (desktop notification and screen flashing) in specified duration.

## Description

time-to-go is a simple command. It accepts a duration to the alarm and when the duration passes, triggers an alarm with desktop notification and screen flashing. It can be used for any miscellaneous purpose, from meditation timer to cooking instant noodle.

## Usage

time-to-go <TIME>

Options:
  -s, --simple
        Simple output which doesn't show remained seconds.
  -h, --help
        Print this help message.
  -v, --version
        Print version information and quit.

Example:

To set alarm 3 minutes 20 seconds.

  $ time-to-go 3:20

time-to-go accepts TIME as the below format. When you specify time unit, it must be one of units defined by International System of Units (SI) or units outside the SI. i.e.) second: s, minute: min

  45 seconds: 45 s, 45s, .45, :45
  3 minutes: 3 min, 3min, 3.00, 3.0, 3. 3:00, 3:0, 3
  2 minutes 40 seconds: 2 min 40 s, 2min 40s, 2 40, 2.40, 2:40
  1 hours 15 minutes: 1 h 15 min, 1h 15min, 1.15.0, 1.15.00, 1:15:0, 1:15:00
  1 hours 20 minutes 30 seconds: 1 h 20 min 30 s, 1h 20min 30s, 1 20 30, 1.20.30, 1:20:30
  2 hours 40 seconds: 2 h 40 s, 2h 40s, 2 0 45

Press Ctrl+C to cancel the timer.

## Install

time-to-go uses Libnotify. So libnotify and Go bindings [go-notify](https://github.com/mqu/go-notify) are required.

### libnotify

RedHat (Please replace by dnf in Fedora)

```bash
$ yum install libnotify-devel
```

Debian/Ubuntu

```bash
$ sudo apt-get install libnotify-dev
```

To install, use `go get`:

```bash
$ go get -d github.com/tdsh/time-to-go
```

## Contribution

1. Fork ([https://github.com/tdsh/time-to-go/fork](https://github.com/tdsh/time-to-go/fork))
1. Create a feature branch
1. Commit your changes
1. Rebase your local changes against the master branch
1. Run test suite with the `go test ./...` command and confirm that it passes
1. Run `gofmt -s`
1. Create a new Pull Request

## Author

[tdsh](https://github.com/tdsh)


CHANGELOG.md
## 0.2.0 (2018-01-16)

### Added

- Accept duration over an hour with various syntax.
- When you specify time unit, it must be one of units defined by International System of Units (SI) or units outside the SI. i.e.) second: s, minute: min, hour: h

## 0.1.0 (2017-08-09)

Initial release

### Added

- Add Fundamental features

### Deprecated

- Nothing

### Removed

- Nothing

### Fixed

- Nothing


cli_test.go
package main

import (
	"bytes"
	"fmt"
	"strings"
	"testing"
)

func TestRun_versionFlag(t *testing.T) {
	outStream, errStream := new(bytes.Buffer), new(bytes.Buffer)
	cli := &CLI{outStream: outStream, errStream: errStream}
	args := strings.Split("./time-to-go -version", " ")

	status := cli.Run(args)
	if status != ExitCodeOK {
		t.Errorf("expected %d to eq %d", status, ExitCodeOK)
	}

	expected := fmt.Sprintf("time-to-go version %s", appVersion)
	if !strings.Contains(errStream.String(), expected) {
		t.Errorf("expected %q to eq %q", errStream.String(), expected)
	}
}


cli.go
package main

import (
	"flag"
	"fmt"
	"io"
	"os"
	"os/signal"
	"strings"
	"sync"
	"time"

	"github.com/mqu/go-notify"
)

// Exit codes are int values that represent an exit code for a particular error.
const (
	ExitCodeOK    int = 0
	ExitCodeError int = 1 + iota
)

// CLI is the command line object
type CLI struct {
	// outStream and errStream are the stdout and stderr
	// to write message from the CLI.
	outStream, errStream io.Writer
}

// Run invokes the CLI with the given arguments.
func (cli *CLI) Run(args []string) int {
	var (
		simple  bool
		version bool
		help    bool
	)

	// Define option flag parse
	flags := flag.NewFlagSet(name, flag.ContinueOnError)
	flags.SetOutput(cli.errStream)

	flags.BoolVar(&simple, "simple", false, "(shortcut: s) Simple output which doesn't show remained seconds.")
	flags.BoolVar(&simple, "s", false, "(shortcut: s) Simple output which doesn't show remained seconds.")
	flags.BoolVar(&version, "version", false, "(shortcut: v) Print version information and quit.")
	flags.BoolVar(&version, "v", false, "(shortcut: v) Print version information and quit.")
	flags.BoolVar(&help, "help", false, "(shortcut: h) Print this message.")
	flags.BoolVar(&help, "h", false, "(shortcut: h) Print this message.")
	flags.Usage = printUsage

	// Parse commandline flag
	if err := flags.Parse(args[1:]); err != nil {
		return ExitCodeError
	}

	if help {
		printUsage()
		return ExitCodeOK
	}

	// Show version
	if version {
		fmt.Fprintf(cli.errStream, "%s version %s\n", name, appVersion)
		return ExitCodeOK
	}

	d, err := getDuration(flags.Args())
	if err != nil {
		fmt.Fprintf(cli.errStream, "\033[31;1m%v\n", err)
		fmt.Fprintf(cli.errStream, "\033[31;1mPlease check usage (%s -h)\033[0m\n", name)
		return ExitCodeError
	}
	rem := int(d.Seconds())

	notify.Init("time-to-go")
	sigCh := make(chan os.Signal)
	signal.Notify(sigCh, os.Interrupt)
	defer close(sigCh)
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()
	stop := make(chan bool)
	defer close(stop)
	fmt.Printf("Sleeping %v\n", strings.Replace(d.String(), "m", "min", 1))
	go func() {
	loop:
		for {
			select {
			case <-ticker.C:
				if simple != true {
					rem--
					if rem != 0 {
						min := rem / 60
						sec := rem % 60
						hour := min / 60
						min = min % 60
						timeLimit := ""
						switch {
						case hour > 0:
							timeLimit = fmt.Sprintf("%02vh%02vmin%02vs", hour, min, sec)
						case min > 0:
							timeLimit = fmt.Sprintf("   %02vmin%02vs", min, sec)
						default:
							timeLimit = fmt.Sprintf("        %02vs", sec)
						}
						fmt.Fprintf(cli.outStream, "\r%s remains...", timeLimit)
					} else {
						fmt.Fprintf(cli.outStream, "\r  0 sec(s) remains...\n")
						break loop
					}
				}
			case <-stop:
				break loop
			}
		}
	}()

	select {
	case <-sigCh:
		fmt.Fprintf(cli.errStream, "\nCancelled.\n")
		return ExitCodeOK
	case <-time.After(d):
		break
	}

	var g sync.WaitGroup
	g.Add(2)
	go func() {
		notify.Init("time-to-go")
		n := notify.NotificationNew("time-to-go", "Wake up!!!!", "appointment-soon")
		n.Show()
		g.Done()
	}()
	go func() {
		flashScreen(6)
		g.Done()
	}()

	g.Wait()
	return ExitCodeOK
}


utils.go
package main

import (
	"errors"
	"flag"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var helpMessage = `Usage:
  time-to-go <TIME>

Options:
  -s, --simple
        Simple output which doesn't show remained seconds.
  -h, --help
        Print this help message.
  -v, --version
        Print version information and quit.

Example:

To set alarm 3 minutes 20 seconds.

  $ time-to-go 3:20

time-to-go accepts TIME as the below format.
When you specify time unit, it must be one of units defined by International System of Units (SI) or units outside the SI.
i.e.) second: s, minute: min, hour: h

  45 seconds: 45 s, 45s, .45, :45
  3 minutes: 3 min, 3min, 3.00, 3.0, 3. 3:00, 3:0, 3:
  2 minutes 40 seconds: 2 min 40 s, 2min 40s, 2 40, 2.40, 2:40
  1 hours 15 minutes: 1 h 15 min, 1h 15min, 1.15.0, 1.15.00, 1:15:0, 1:15:00
  1 hours 20 minutes 30 seconds: 1 h 20 min 30 s, 1h 20min 30s, 1 20 30, 1.20.30, 1:20:30
  2 hours 40 seconds: 2 h 40 s, 2h 40s, 2 0 45

Press Ctrl+C to cancel the timer.
`

var re = regexp.MustCompile(`:+`)

// printUsage prints help message.
func printUsage() {
	fmt.Fprintf(os.Stderr, helpMessage)
	flag.PrintDefaults()
}

// getDuration converts args to time.Duration.
// It accepts various format. If conversion fails, corresponding error
// is returned.
func getDuration(args []string) (time.Duration, error) {
	var d time.Duration
	argString := strings.Join(args, ":")

	// 1. Replace "." with ":".
	argString = strings.Replace(argString, ".", ":", -1)
	// 2. Check the exisistence of "h", "min" and "s".
	if strings.Contains(argString, "h") && !strings.Contains(argString, "min") && strings.Contains(argString, "s") {
		argString = strings.Replace(argString, "h:", "h:0:", 1)
	}
	// 2. Replace "h" with ":".
	argString = strings.Replace(argString, "h", ":", 1)
	// 3. Replace "min" with ":".
	argString = strings.Replace(argString, "min", ":", 1)
	// 4. Replace "::" with ":".
	argString = re.ReplaceAllString(argString, ":")
	// 5. Trim appended "s".
	if strings.HasSuffix(argString, ":s") {
		argString = strings.TrimRight(argString, ":s")
	}
	if strings.HasSuffix(argString, "s") {
		argString = strings.TrimRight(argString, "s")
	}
	// 5. Prepend ":" if it consists of numbers only.
	_, err := strconv.Atoi(argString)
	if err == nil {
		argString = ":" + argString
	}

	err = nil
	array := strings.Split(argString, ":")
	switch len(array) {
	case 2:
		switch {
		case array[0] == "":
			s, err := strconv.Atoi(array[1])
			if err != nil {
				break
			}
			d = time.Duration(s) * time.Second
		case array[1] == "":
			m, err := strconv.Atoi(array[0])
			if err != nil {
				break
			}
			d = time.Duration(m) * time.Minute
		default:
			m, err := strconv.Atoi(array[0])
			if err != nil {
				break
			}
			s, err := strconv.Atoi(array[1])
			if err != nil {
				break
			}
			d = time.Duration(m)*time.Minute + time.Duration(s)*time.Second
		}
	case 3:
		switch {
		case array[2] == "":
			h, err := strconv.Atoi(array[0])
			if err != nil {
				break
			}
			m, err := strconv.Atoi(array[1])
			if err != nil {
				break
			}
			d = time.Duration(h)*time.Hour + time.Duration(m)*time.Minute
		default:
			h, err := strconv.Atoi(array[0])
			if err != nil {
				break
			}
			m, err := strconv.Atoi(array[1])
			if err != nil {
				break
			}
			s, err := strconv.Atoi(array[2])
			if err != nil {
				break
			}
			d = time.Duration(h)*time.Hour + time.Duration(m)*time.Minute + time.Duration(s)*time.Second
		}
	default:
		err = errors.New("Wrong format")
	}
	return d, err
}

// flashScreen makes current terminal screen flashing for t times
// using ANSI escape sequences. It has the potential not to work in
// some terminal environments.
func flashScreen(t int) {
	// workaround for TMUX.
	escSeq := "\x1b[?%s"
	_, tmux := os.LookupEnv("TMUX")
	if tmux == true {
		escSeq = "\x1bPtmux;\x1b\x1b[?%s\x1b\\"
	}
	for i := 0; i < t; i++ {
		// reverse video
		fmt.Fprintf(os.Stdout, escSeq, "5h")
		time.Sleep(500 * time.Millisecond)
		// normal video
		fmt.Fprintf(os.Stdout, escSeq, "5l")
		if i == t-1 {
			break
		}
		time.Sleep(500 * time.Millisecond)
	}
}
